File: server/app/models/position.py
from dataclasses import dataclass

@dataclass(frozen=True)
class Position:
    """Represents a position on the chess board (row, col)"""
    row: int
    col: int
    
    def to_algebraic(self) -> str:
        """Convert to chess notation (e.g., e4)"""
        file = chr(ord('a') + self.col)
        rank = str(8 - self.row)
        return f"{file}{rank}"
    
    @classmethod
    def from_algebraic(cls, algebraic: str) -> 'Position':
        """Create Position from chess notation (e.g., 'e4')"""
        if len(algebraic) != 2:
            raise ValueError(f"Invalid algebraic notation: {algebraic}")
        
        col = ord(algebraic[0].lower()) - ord('a')
        row = 8 - int(algebraic[1])
        return cls(row, col)
    
    def is_valid(self) -> bool:
        """Check if position is within board bounds"""
        return 0 <= self.row < 8 and 0 <= self.col < 8
    
    def offset(self, row_delta: int, col_delta: int) -> 'Position':
        """Get a new position offset by the given deltas"""
        return Position(self.row + row_delta, self.col + col_delta)
    
    def __str__(self) -> str:
        return self.to_algebraic()
    
    def __repr__(self) -> str:
        return f"Position({self.row}, {self.col})"

File: client/src/game/position.ts
import type { PositionData } from "../types/chess_types";

export class Position {
  row: number;
  col: number;

  constructor(row: number, col: number) {
    this.row = row;
    this.col = col;
  }

  isValid(): boolean {
    return this.row >= 0 && this.row < 8 && this.col >= 0 && this.col < 8;
  }

  toAlgebraic(): string {
    const file = String.fromCharCode("a".charCodeAt(0) + this.col);
    const rank = String(8 - this.row);
    return `${file}${rank}`;
  }

  static fromAlgebraic(algebraic: string): Position {
    if (algebraic.length !== 2) {
      throw new Error(`Invalid algebraic notation: ${algebraic}`);
    }

    const file = algebraic[0].toLowerCase();
    const rank = algebraic[1];

    const col = file.charCodeAt(0) - "a".charCodeAt(0);
    const row = 8 - Number(rank);

    return new Position(row, col);
  }

  offset(rowDelta: number, colDelta: number): Position {
    return new Position(this.row + rowDelta, this.col + colDelta);
  }

  toString(): string {
    return this.toAlgebraic();
  }

  serializePosition(): PositionData {
    return { row: this.row, col: this.col };
  }

  repr(): string {
    return `Position(${this.row}, ${this.col})`;
  }
}
